src/lib/coda_base/transaction_logic.ml

```diff
@@ -374,112 +401,267 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
       let%bind () = validate_nonces nonce account.nonce in
-      (* TODO: Put actual value here. See issue #4036. *)
-      let current_global_slot = Global_slot.zero in
-      let%bind () =
-        validate_time ~valid_until:payload.common.valid_until
-          ~current_global_slot
+      let undo_common : Undo.User_command_undo.Common.t =
+        {user_command; previous_receipt_chain_hash= account.receipt_chain_hash}
       in
       let%map timing =
-        if User_command.Payload.is_payment payload then
-          let txn_amount =
-            match User_command.Payload.body payload with
-            | Payment {amount; _} ->
-                amount
-            | _ ->
-                failwith "Expected payment when validating transaction timing"
-          in
-          validate_timing ~txn_amount ~txn_global_slot:current_global_slot
-            ~account
-        else return account.timing
+        validate_timing ~txn_amount:fee ~txn_global_slot:current_global_slot
+          ~account
       in
-      let account =
-        { account with
-          nonce= Account.Nonce.succ account.nonce
+      ( location
+      , { account with
+          balance
+        ; nonce= Account.Nonce.succ account.nonce
         ; receipt_chain_hash=
             Receipt.Chain_hash.cons payload account.receipt_chain_hash
         ; timing }
-      in
-      ({account with balance}, common)
+      , undo_common )
     in
-    match User_command.Payload.body payload with
-    | Stake_delegation (Set_delegate {new_delegate}) ->
-        (*new delegate should be in the ledger*)
-        (*TODO transaction snark doesn't check this*)
-        let%bind delegate_location = location_of_key' ledger "" new_delegate in
-        let%map _ = get' ledger "new_delegate" delegate_location in
-        set ledger sender_location {sender_account with delegate= new_delegate} ;
-        { Undo.User_command_undo.common
-        ; body= Stake_delegation {previous_delegate= sender_account.delegate}
-        }
-    | Payment Payment_payload.Poly.{amount; receiver} ->
-        let%bind sender_balance' = sub_amount sender_account.balance amount in
-        let undo emptys : Undo.User_command_undo.t =
-          {common; body= Payment {previous_empty_accounts= emptys}}
-        in
-        if Public_key.Compressed.equal sender receiver then (
-          ignore @@ set ledger sender_location sender_account ;
-          return (undo []) )
+    (* Charge the fee. This must happen, whether or not the command itself
+       succeeds, to ensure that the network is compensated for processing this
+       command.
+    *)
+    set_with_location ledger fee_payer_location fee_payer_account ;
+    let source = User_command.source user_command in
+    let receiver = User_command.receiver user_command in
+    (* Compute the necessary changes to apply the command, failing if any of
+       the conditions are not met.
+    *)
+    let exception Reject of Error.t in
+    match
```

Comment:
Probably done so to make the implementations match up better, but I also think this long nested match expression is difficult to parse.